src/coxph.jl

```diff
@@ -0,0 +1,85 @@
+function _cox_f(β, c::CoxAux{T})::T where T
+    grad = zeros(T, length(β))
+    hes = zeros(T, length(β), length(β))
+    _cox_fgh!(β, grad,hes, c::CoxAux{T}, false)
+end
+
+function _cox_fgh!(β, grad,hes, c::CoxAux{T}, compute_derivatives)::T where T
+    #get relevant quantities to compute negative loglikelihood, gradient and hessian
+
+    update_cox!(c, β, compute_derivatives)
+
+    X, ξ, Xβ, θ, Xθ, ξθ, λ, fs, ls  =
+        c.X, c.ξ, c.Xβ, c.θ, c.Xθ, c.ξθ, c.λ, c.fs, c.ls
+
+
+    #compute negative loglikelihood, gradient and hessian
+    y = zero(T)
+
+    if compute_derivatives
+        grad .= zero(T)
+        hes .= zero(T)
+
+        # preallocate
+        Z = zeros(T, length(β))
+    end
+
+    @inbounds for i in 1:length(fs)
+        for j in (fs[i]):(ls[i])
+            ρ = (j-fs[i])/(ls[i] + 1 - fs[i])
+            ϕ = θ.tails[i]-ρ * θ.chunks[i]
+            y -= Xβ[j] -log(ϕ)
+            if compute_derivatives
+                for k in eachindex(Z)
+                    Z[k] = Xθ.tails[i, k] - ρ * Xθ.chunks[i, k]
+                end
+                for k2 in 1:length(β)
+                    grad[k2] += Z[k2]/ϕ - X[j,k2]
+                    for k1 in 1:k2
+                        Ξ = ξθ.tails[i, k1, k2] - ρ * ξθ.chunks[i, k1, k2]
+                        hes[k1, k2] += Ξ/ϕ - Z[k1] * Z[k2]/ϕ^2
+                    end
+                end
+            end
+        end
+    end
+    y += λ*(β' * β)
```

Comment:
At least on 0.6, I see the opposite:
```
julia> x = rand(10_000);

julia> @benchmark $x'*$x
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     6.333 μs (0.00% GC)
  median time:      7.041 μs (0.00% GC)
  mean time:        7.718 μs (0.00% GC)
  maximum time:     39.177 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     5

julia> @benchmark dot($x, $x)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     6.545 μs (0.00% GC)
  median time:      6.647 μs (0.00% GC)
  mean time:        6.931 μs (0.00% GC)
  maximum time:     20.233 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     5

julia> @benchmark sum(i->i^2, $x)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     2.808 μs (0.00% GC)
  median time:      2.831 μs (0.00% GC)
  mean time:        2.912 μs (0.00% GC)
  maximum time:     10.364 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     9
```
Not surprising that the `x'x` form is equivalent in speed to `dot(x, x)`, since `@code_lowered` shows that the former is calling dot. TIL. 🙂 

I tried to benchmark on master too but BenchmarkTools seems completely broken on master.